File: src/algorithms/hash-maps/mapping/roman-to-integer/sources/roman-to-integer.ts
// Roman to Integer — convert a Roman numeral string to its integer value using a lookup map
function romanToInteger(text: string): number {
  const romanMap = new Map<string, number>(); // @step:initialize
  romanMap.set("I", 1); // @step:insert-key
  romanMap.set("V", 5); // @step:insert-key
  romanMap.set("X", 10); // @step:insert-key
  romanMap.set("L", 50); // @step:insert-key
  romanMap.set("C", 100); // @step:insert-key
  romanMap.set("D", 500); // @step:insert-key
  romanMap.set("M", 1000); // @step:insert-key
  let totalValue = 0;
  for (let charIndex = 0; charIndex < text.length; charIndex++) {
    const currentSymbol = text[charIndex]!; // @step:lookup-key
    const nextSymbol = text[charIndex + 1];
    const currentValue = romanMap.get(currentSymbol)!; // @step:key-found
    const nextValue = nextSymbol !== undefined ? (romanMap.get(nextSymbol) ?? 0) : 0;
    if (currentValue < nextValue) {
      totalValue -= currentValue; // @step:key-found
    } else {
      totalValue += currentValue; // @step:key-found
    }
  }
  return totalValue; // @step:complete
}

File: src/algorithms/hash-maps/mapping/roman-to-integer/sources/roman-to-integer.py
# Roman to Integer — convert a Roman numeral string to its integer value using a lookup map
def roman_to_integer(text):
    roman_map = {}  # @step:initialize
    roman_map["I"] = 1  # @step:insert-key
    roman_map["V"] = 5  # @step:insert-key
    roman_map["X"] = 10  # @step:insert-key
    roman_map["L"] = 50  # @step:insert-key
    roman_map["C"] = 100  # @step:insert-key
    roman_map["D"] = 500  # @step:insert-key
    roman_map["M"] = 1000  # @step:insert-key
    total_value = 0
    for char_index in range(len(text)):
        current_symbol = text[char_index]  # @step:lookup-key
        current_value = roman_map[current_symbol]  # @step:key-found
        next_value = roman_map[text[char_index + 1]] if char_index + 1 < len(text) else 0
        if current_value < next_value:
            total_value -= current_value  # @step:key-found
        else:
            total_value += current_value  # @step:key-found
    return total_value  # @step:complete
